sql: output RU estimate for EXPLAIN ANALYZE on tenants

DrewKimball opened the pull request:
**sql: surface query request units consumed by network egress**
This commit adds a top-level field to the output of `EXPLAIN ANALYZE`
that shows the estimated number of RUs that would be consumed due to
network egress to the client. The estimate is obtained by measuring
the in-memory size of the query result, and passing that to the
tenant cost config's `PGWireEgressCost` method.

**sql: surface query request units consumed due to cpu usage**
This commit adds the ability for clients to estimate the number of RUs
consumed by a query due to CPU usage. This is accomplished by keeping a
moving average of the CPU usage for the entire tenant process, then using
that to obtain an estimate for what the CPU usage *would* be if the query
wasn't running. This is then compared against the actual measured CPU usage
during the query's execution to get the estimate. For local flows this is
done at the `connExecutor` level; for remote flows this is handled by the
last outbox on the node (which gathers and sends the flow's metadata).
The resulting RU estimate is added to the existing estimate from network
egress and displayed in the output of `EXPLAIN ANALYZE`.

**sql: surface query request units consumed by IO**
This commit adds tracking for request units consumed by IO operations
for all execution operators that perform KV operations. The corresponding
RU count is recorded in the span and later aggregated with the RU consumption
due to network egress and CPU usage. The resulting query RU consumption
estimate is visible in the output of `EXPLAIN ANALYZE`.

**multitenantccl: add sanity testing for ru estimation**
This commit adds a sanity test for the RU estimates produced by running
queries with `EXPLAIN ANALYZE` on a tenant. The test runs each test query
several times with `EXPLAIN ANALYZE`, then runs all test queries without
`EXPLAIN ANALYZE` and compares the resulting actual RU measurement to the
aggregated estimates.

Informs #74441

Release note (sql change): Added an estimate for the number of request units
consumed by a query to the output of `EXPLAIN ANALYZE` for tenant sessions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89256)
<!-- Reviewable:end -->

andy-kimball commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go:
```diff
@@ -790,6 +792,13 @@ func (c *tenantSideCostController) OnResponseWait(
 	}
 
+	// Record the number of RUs consumed by the IO request.
+	if sp := tracing.SpanFromContext(ctx); sp != nil && sp.RecordingType() != tracingpb.RecordingOff {
```
This should go above the lock, since we want to release the lock as quickly as possible.

andy-kimball commented on pkg/sql/distsql_running.go:
```diff
@@ -1172,6 +1173,7 @@ func (r *DistSQLReceiver) Push(
 	if r.resultWriter.Err() == nil && r.ctx.Err() != nil {
 		r.SetError(r.ctx.Err())
 	}
+	r.stats.networkEgressEstimate += int64(row.Size())
```
Is this or the statement below expensive to calculate? If we do something like looping that might be slow for large rows/batches, we might consider only doing it when it's needed. That said, it'd be more complex, so we should only do that if it's a real issue.

DrewKimball replied:
Yeah, I'll want to hide this all behind some setting so we only do it for tenants that run explain analyze, since overhead can be non-negligible

DrewKimball commented:
TODO: squash commits before merging.

DrewKimball commented:
Maybe hold off on reviewing for a bit until I can get all the tests passing.

DrewKimball commented:
Alright, the only test failures I see now are this:
```
    lint_test.go:91: 
        pkg/server/stop_trigger.go:46:3: github.com/cockroachdb/cockroach/pkg/util/log.Infof format %s has arg r of wrong type github.com/cockroachdb/cockroach/pkg/server.ShutdownRequest
    --- FAIL: TestLint/TestRoachVet (144.88s)
```
and this:
```
[17:46:43][Run unit tests] FAIL: //pkg/kv/kvclient/kvcoord:kvcoord_disallowed_imports_test (see /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvclient/kvcoord/kvcoord_disallowed_imports_test/test.log)
[17:46:43][Run unit tests] INFO: From Testing //pkg/kv/kvclient/kvcoord:kvcoord_disallowed_imports_test:
[17:46:43][Run unit tests] ==================== Test output for //pkg/kv/kvclient/kvcoord:kvcoord_disallowed_imports_test:
[17:46:43][Run unit tests] ERROR: //pkg/kv/kvclient/kvcoord:kvcoord imports //pkg/storage:storage
[17:46:43][Run unit tests] 	check: bazel query 'somepath(//pkg/kv/kvclient/kvcoord:kvcoord, //pkg/storage:storage)'
[17:46:43][Run unit tests] ================================================================================
```
both of which seem unrelated.

yuzefovich reviewed (commented):
Nice work!

> Reviewed 1 of 13 files at r1, 1 of 16 files at r2, 5 of 27 files at r6, 24 of 24 files at r13, 18 of 18 files at r14, 16 of 16 files at r15, 3 of 3 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-:-NFVPdACDDp197M1B9ak:bdtx8r6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 878 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVJJ43-XCWljFEZPFi:-NFVJJ43-XCWljFEZPFj:b-w0aflb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L878)):*
> ```Go
> 	c.mu.Lock()
> 	cpuAvg = c.mu.avgCPUPerSec
> 	c.mu.Unlock()
> ```

nit: I believe in one of the recent Go releases the price of the `defer` has gone done significantly, so we switched to mostly `Unlock`ing in a `defer` in situations like this, even in performance sensitive code. But up to you.
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 9 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVKAUr8e-rqhsgRveN:-NFVKAUr8e-rqhsgRveO:buabxao) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/multitenant/multitenantcpu/cpu_usage.go#L9)):*
> ```Go
> //     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
> 
> package multitenantcpu
> ```

Should this be somewhere in `pkg/ccl` subtree? Alternatively, should this use BSL license (similar to `multitenantio`)?
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 63 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVLt8NEYG0ZhbNcZR1:-NFVLt8NEYG0ZhbNcZR2:b408bwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/multitenant/multitenantcpu/cpu_usage.go#L63)):*
> ```Go
> // It is used for measuring tenant RU consumption.
> func GetCPUSeconds(ctx context.Context) (cpuSecs float64) {
> 	userTimeMillis, sysTimeMillis, err := status.GetCPUTime(ctx)
> ```

`GetCPUTime` doesn't actually use `ctx` argument. We could remove it which will reduce the amount of plumbing needed here.
___
*[`pkg/sql/distsql_running.go` line 617 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVGkxOBXNuLnMpO2Mv:-NFVGkxOBXNuLnMpO2Mw:b-qjevjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f28983582f1d8697ec6bd94b923a16885d257f30/pkg/sql/distsql_running.go#L617)):*
> ```Go
> 			// Only collect the network egress estimate for a tenant that is running
> 			// EXPLAIN ANALYZE, since the overhead is non-negligible.
> 			recv.isTenantExplainAnalyze = instrumentation.collectExecStats &&
> ```

nit: I think the second part of the AND is sufficient, no?
___
*[`pkg/sql/colflow/stats.go` line 234 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVNig40FfXU8wcqoRR:-NFVNig40FfXU8wcqoRS:bypcv47) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/colflow/stats.go#L234)):*
> ```Go
> 		execstats.PopulateKVMVCCStats(&s.KV, &scanStats)
> 		if vsc.isTenant {
> 			// Only set the RUs consumed if this is a tenant.
> ```

nit: would it be ok to set it for single tenant deployments too (since `RuConsumed` would be zero)? This would then be ignored in `instrumentation.go`.
___
*[`pkg/sql/colflow/vectorized_flow.go` line 506 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVMeXy9PdJpBe2gZuy:-NFVMeXy9PdJpBe2gZuz:b6cujov) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/sql/colflow/vectorized_flow.go#L506)):*
> ```Go
> 					MaxMemUsage:  optional.MakeUint(uint64(flowCtx.Mon.MaximumBytes())),
> 					MaxDiskUsage: optional.MakeUint(uint64(flowCtx.DiskMonitor.MaximumBytes())),
> 					ConsumedRU:   optional.MakeUint(uint64(cpuStatsCollector.EndCollection(ctx))),
> ```

We also need to make a similar change to `flowinfra/outbox.go` for the row-based flows.
___
*[`pkg/sql/colflow/vectorized_flow.go` line 750 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVLj-D4h8BykoCzIzl:-NFVLj-D4h8BykoCzIzm:bxxc1ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/sql/colflow/vectorized_flow.go#L750)):*
> ```Go
> 	stream *execinfrapb.StreamEndpointSpec,
> 	factory coldata.ColumnFactory,
> 	getStats func(ctx2 context.Context) []*execinfrapb.ComponentStats,
> ```

nit: can omit the variable name altogether.
___
*[`pkg/sql/execstats/stats.go` line 77 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVNwSm3rOWJM-P-SNg:-NFVNwSn5iBZw1TeE32c:bn8gpnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/execstats/stats.go#L77)):*
> ```Go
> 	// RuConsumed is the number of RUs that were consumed during the course of a
> 	// scan.
> 	RuConsumed uint64
> ```

nit: maybe `s/RuConsumed/ConsumedRU/` for symmetry with the proto?
___
*[`pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go` line 67 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVOf5g2lOkuxC-YWKx:-NFVOf5g2lOkuxC-YWKy:bkwlozp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go#L67)):*
> ```Go
> 	}
> 
> 	params.DisableDefaultTestTenant = true
> ```

nit: this is already set above.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Very nice! I'm not so familiar with this code so someone else should approve, but I just had a couple of small questions.

> Reviewed 1 of 13 files at r1, 1 of 16 files at r2, 5 of 27 files at r6, 24 of 24 files at r13, 18 of 18 files at r14, 16 of 16 files at r15, 3 of 3 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-:-NFp1k39DxAu_5cLWhl8:b-ibz0a1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, and @yuzefovich)

___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 52 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFp0GxU2_IVp1g_GcGI:-NFp0GxU2_IVp1g_GcGJ:b972xrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/multitenant/multitenantcpu/cpu_usage.go#L52)):*
> ```Go
> 	cpuDelta := GetCPUSeconds(ctx) - h.startCPU
> 	timeElapsed := timeutil.Since(h.startTime)
> 	expectedCPUDelta := timeElapsed.Seconds() * h.avgCPUPerSecond
> ```

I guess this won't work well when there are many concurrent queries?
___
*[`pkg/sql/rowexec/inverted_joiner.go` line 782 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFoyFYn8SqPvEniPjy0:-NFoyFYo0KrUut8l-njU:beo6sze) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/rowexec/inverted_joiner.go#L782)):*
> ```Go
> 	}
> 	if ij.FlowCtx.Cfg.TenantCostController != nil {
> 		// Don't populate the consumed RU unless this is a tenant.
> ```

Would it be bad to collect this for non-tenants? I know someone was asking if it would be possible to surface RU estimates even for non-tenants so users could know what their workload might cost on CRDB Serverless. Then you could also avoid a lot of plumbing.

(maybe the same question @yuzefovich had)
___
*[`pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go` line 43 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFp-5ws6yGP8lDY5Zw_:-NFp-5ws6yGP8lDY5Zwa:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go#L43)):*
> ```Go
> // consumption is within reasonable distance from the actual measured RUs for
> // the tenant.
> func TestEstimateQueryRUConsumption(t *testing.T) {
> ```

Nice test!
___
*[`pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go` line 136 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFozSqQ8vHlmsiz5GzH:-NFozSqQ8vHlmsiz5GzI:bi7d5qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go#L136)):*
> ```Go
> 				if strings.Contains(val, "estimated RUs consumed") {
> 					substr := strings.Split(val, " ")
> 					if len(substr) == 4 {
> ```

if this is false should you return an error?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @rytaft, and @yuzefovich)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 796 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NDVPx2AUJAOY1pX_sNB-r1-796:-NFVmSxs9yrt2KrffJv_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/888cf4178e08ced53bc511ce993653d018909043/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L796)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This should go above the lock, since we want to release the lock as quickly as possible.
</blockquote></details>

Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 878 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVJJ43-XCWljFEZPFi:-NFVZtTB8c_kAji1e8xc:b-lsc9q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L878)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I believe in one of the recent Go releases the price of the `defer` has gone done significantly, so we switched to mostly `Unlock`ing in a `defer` in situations like this, even in performance sensitive code. But up to you.
</blockquote></details>

Done. That's good to know
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 9 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVKAUr8e-rqhsgRveN:-NFV_71h6zNMkurPdECt:b-h8bg43) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/multitenant/multitenantcpu/cpu_usage.go#L9)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Should this be somewhere in `pkg/ccl` subtree? Alternatively, should this use BSL license (similar to `multitenantio`)?
</blockquote></details>

Oops, just copied that license blindly. I'm not really sure of where things are supposed to be located WRT `multitenant` vs `multitenantccl` but for now I've just changed the license. Done.
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 63 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVLt8NEYG0ZhbNcZR1:-NFVkVgNEFIfVJw2Ig65:b-6g9iaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/multitenant/multitenantcpu/cpu_usage.go#L63)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`GetCPUTime` doesn't actually use `ctx` argument. We could remove it which will reduce the amount of plumbing needed here.
</blockquote></details>

That's a good point. How would you handle the logging here without the context though? Returning the error would just run into the same problem further up the stack. I guess an option would be to just return 0 on error, but that seems problematic.
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 52 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFp0GxU2_IVp1g_GcGI:-NGIWJaGBOqFmGLbjZJH:b114b51) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/multitenant/multitenantcpu/cpu_usage.go#L52)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I guess this won't work well when there are many concurrent queries?
</blockquote></details>

Yes, I'll add a comment. Done.
___
*[`pkg/sql/distsql_running.go` line 617 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVGkxOBXNuLnMpO2Mv:-NFVkx_00V5qvssNSAhc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f28983582f1d8697ec6bd94b923a16885d257f30/pkg/sql/distsql_running.go#L617)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think the second part of the AND is sufficient, no?
</blockquote></details>

Done.
___
*[`pkg/sql/colflow/stats.go` line 234 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVNig40FfXU8wcqoRR:-NFVlu_mBdU3UqF5goiq:bkpcxtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/colflow/stats.go#L234)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: would it be ok to set it for single tenant deployments too (since `RuConsumed` would be zero)? This would then be ignored in `instrumentation.go`.
</blockquote></details>

Good point, that cleans up some plumbing I had to do. Done.
___
*[`pkg/sql/colflow/vectorized_flow.go` line 506 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVMeXy9PdJpBe2gZuy:-NGO7iaS6xnzA1YH9epT:bbuk5o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/sql/colflow/vectorized_flow.go#L506)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We also need to make a similar change to `flowinfra/outbox.go` for the row-based flows.
</blockquote></details>

Done. I've moved the cpu usage collector for remote flows to the `FlowCtx` since I didn't see a good way to propagate it to the row-based outbox otherwise.
___
*[`pkg/sql/colflow/vectorized_flow.go` line 750 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVLj-D4h8BykoCzIzl:-NFVh2M67ZbVNKHZCoJ1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/sql/colflow/vectorized_flow.go#L750)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: can omit the variable name altogether.
</blockquote></details>

Done.
___
*[`pkg/sql/execstats/stats.go` line 77 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVNwSm3rOWJM-P-SNg:-NFVmHhj9RyEVYeLHIwq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/execstats/stats.go#L77)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: maybe `s/RuConsumed/ConsumedRU/` for symmetry with the proto?
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/inverted_joiner.go` line 782 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFoyFYn8SqPvEniPjy0:-NGIVlOn4GZmU_8x1JKK:beu4u9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/rowexec/inverted_joiner.go#L782)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Would it be bad to collect this for non-tenants? I know someone was asking if it would be possible to surface RU estimates even for non-tenants so users could know what their workload might cost on CRDB Serverless. Then you could also avoid a lot of plumbing.

(maybe the same question @yuzefovich had)
</blockquote></details>

Good point, Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go` line 43 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFp-5ws6yGP8lDY5Zw_:-NGIX22J1Uk8n8ynpTCU:b-bgb23r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go#L43)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Nice test!
</blockquote></details>

Thanks! Took some doing to figure out how to make it run.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go` line 67 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVOf5g2lOkuxC-YWKx:-NFVmMmVCxSMAwn67GET:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go#L67)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this is already set above.
</blockquote></details>

Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go` line 136 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFozSqQ8vHlmsiz5GzH:-NGIW4fnE4F4vGFc_-Ad:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c788b957023e589fa5db77189cd0d2e166171fa/pkg/ccl/multitenantccl/tenantcostclient/query_ru_estimate_test.go#L136)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

if this is false should you return an error?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 16 files at r2, 31 of 31 files at r17, 21 of 21 files at r18, 14 of 14 files at r19, 5 of 5 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-:-NGOeBhh0YNGkC-trzHK:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, and @rytaft)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 878 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOcmkhBFhF6HyQXkXn:-NGOcmkhBFhF6HyQXkXo:b-386a80) ([raw file](https://github.com/cockroachdb/cockroach/blob/943be1ac2e7ea28a4ed9639e2e1df6df4e2d8351/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L878)):*
> ```Go
> 	c.mu.Lock()
> 	defer c.mu.Unlock()
> 	cpuAvg = c.mu.avgCPUPerSec
> ```

nit: could remove the named return argument and just return the mutex-protected thing.
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 63 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVLt8NEYG0ZhbNcZR1:-NGOaoMc5cEuwB3LWCCP:b-lxcxji) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/multitenant/multitenantcpu/cpu_usage.go#L63)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

That's a good point. How would you handle the logging here without the context though? Returning the error would just run into the same problem further up the stack. I guess an option would be to just return 0 on error, but that seems problematic.
</blockquote></details>

Right, I didn't realize that we also use `ctx` in the current function, nvm.
___
*[`pkg/sql/conn_executor.go` line 16 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOavIiAXeAD2Ajxdb5:-NGOavIiAXeAD2Ajxdb6:b-rujyof) ([raw file](https://github.com/cockroachdb/cockroach/blob/943be1ac2e7ea28a4ed9639e2e1df6df4e2d8351/pkg/sql/conn_executor.go#L16)):*
> ```Go
> 	"context"
> 	"fmt"
> 	"github.com/cockroachdb/cockroach/pkg/multitenant/multitenantcpu"
> ```

nit: the ordering is off in the second commit.
___
*[`pkg/sql/execinfra/flow_context.go` line 17 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOb9m6-8rMICA2czWc:-NGOb9m6-8rMICA2czWd:b-vu7b33) ([raw file](https://github.com/cockroachdb/cockroach/blob/943be1ac2e7ea28a4ed9639e2e1df6df4e2d8351/pkg/sql/execinfra/flow_context.go#L17)):*
> ```Go
> import (
> 	"context"
> 	"github.com/cockroachdb/cockroach/pkg/multitenant/multitenantcpu"
> ```

nit: ditto for the second commit.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @rytaft)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 406 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVJHScCDj2q0lYTd1C:-NFVJHSfAuF3YOY8wTdP:b-yrwo6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/227c501f4c79d4efc4119e01b4b9b9a9a9583ca6/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L406)):*
> ```Go
> 		// Keep track of an exponential moving average of CPU usage.
> 		avg := deltaCPU / deltaTime.Seconds()
> 		c.mu.Lock()
> ```

I'd move this code down into the lock that we're already taking, so that we don't have to acquire/release twice.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 871 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFnhNnICkmimdXH7qMV:-NFnhNnJ9AC-MAqO7iwr:bupiho1) ([raw file](https://github.com/cockroachdb/cockroach/blob/227c501f4c79d4efc4119e01b4b9b9a9a9583ca6/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L871)):*
> ```Go
> 	}
> 
> 	c.mu.Lock()
> ```

Don't we need to add RU here as well? Could it show up in an `EXPLAIN ANALYZE` of an EXPORT or some other statement?
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 60 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOmLvE0JoSAQRVs-fF:-NGOmLvFC2kvD3owjWi8:b1ap6tr) ([raw file](https://github.com/cockroachdb/cockroach/blob/783dcb77a8cc54b35cbac9a905e98147d7008203/pkg/multitenant/multitenantcpu/cpu_usage.go#L60)):*
> ```Go
> 	expectedCPUDelta := timeElapsed.Seconds() * h.avgCPUPerSecond
> 	cpuUsageSeconds := cpuDelta - expectedCPUDelta
> 	if cpuUsageSeconds < 0 {
> ```

I think we should also bound the consumed CPU on the upper side by checking if `cpuUsageSeconds` exceeds `timeElapsed.Seconds()`. I don't believe that we do much parallel CPU work for a single query, and so if it's greater, it's almost always going to be due to other queries interfering with the CPU measurement. I'd rather bound it to reduce that possibility of gross error.
___
*[`pkg/sql/rowexec/inverted_joiner.go` line 782 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFoyFYn8SqPvEniPjy0:-NGOnLHC14FDSDsMYqXC:bsl39pq) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/rowexec/inverted_joiner.go#L782)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Good point, Done.
</blockquote></details>

Note that we don't calculate RU for non-tenants today, so `ConsumedRU` should be zero in those cases.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @rytaft, and @yuzefovich)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 406 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVJHScCDj2q0lYTd1C:-NGQVtd81k4-ZPu68UyO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/227c501f4c79d4efc4119e01b4b9b9a9a9583ca6/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L406)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd move this code down into the lock that we're already taking, so that we don't have to acquire/release twice.
</blockquote></details>

Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 871 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFnhNnICkmimdXH7qMV:-NGQX1RH3TTnzg5VqWK8:b9vqxmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/227c501f4c79d4efc4119e01b4b9b9a9a9583ca6/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L871)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Don't we need to add RU here as well? Could it show up in an `EXPLAIN ANALYZE` of an EXPORT or some other statement?
</blockquote></details>

I added a TODO for that above, but you make a good point that we probably want to handle such cases now. It's not obvious to me how to propagate this information for a query. Do you think it would be ok to just disable RU consumption estimation for "non-simple" statements?
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 878 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOcmkhBFhF6HyQXkXn:-NGQTOJv91KbmA35mkjp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/943be1ac2e7ea28a4ed9639e2e1df6df4e2d8351/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L878)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: could remove the named return argument and just return the mutex-protected thing.
</blockquote></details>

Done.
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 63 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFVLt8NEYG0ZhbNcZR1:-NGQTPfS5g1hUMsvPFNh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d068150fa1e148a8f610a8c479126ffdb999281/pkg/multitenant/multitenantcpu/cpu_usage.go#L63)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Right, I didn't realize that we also use `ctx` in the current function, nvm.
</blockquote></details>

Done.
___
*[`pkg/multitenant/multitenantcpu/cpu_usage.go` line 60 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOmLvE0JoSAQRVs-fF:-NGQUG24BF03oKIX2In9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/783dcb77a8cc54b35cbac9a905e98147d7008203/pkg/multitenant/multitenantcpu/cpu_usage.go#L60)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we should also bound the consumed CPU on the upper side by checking if `cpuUsageSeconds` exceeds `timeElapsed.Seconds()`. I don't believe that we do much parallel CPU work for a single query, and so if it's greater, it's almost always going to be due to other queries interfering with the CPU measurement. I'd rather bound it to reduce that possibility of gross error.
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor.go` line 16 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOavIiAXeAD2Ajxdb5:-NGQaGUrAXarfMbZlHe9:b-7d2pw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/943be1ac2e7ea28a4ed9639e2e1df6df4e2d8351/pkg/sql/conn_executor.go#L16)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the ordering is off in the second commit.
</blockquote></details>

Ah, forgot I haven't set up crlfmt for goland on this laptop yet. Done.
___
*[`pkg/sql/execinfra/flow_context.go` line 17 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NGOb9m6-8rMICA2czWc:-NGQaMwF6IEHmqmt0_SP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/943be1ac2e7ea28a4ed9639e2e1df6df4e2d8351/pkg/sql/execinfra/flow_context.go#L17)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto for the second commit.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 31 of 31 files at r17, 12 of 21 files at r18, 16 of 16 files at r21, 10 of 11 files at r22, 2 of 2 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-:-NGS4xaN7XVk6dL-v7mX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @DrewKimball, and @yuzefovich)



<!-- Sent from Reviewable.io -->

DrewKimball committed:
**sql: surface query request units consumed by network egress**

This commit adds a top-level field to the output of `EXPLAIN ANALYZE`
that shows the estimated number of RUs that would be consumed due to
network egress to the client. The estimate is obtained by buffering
each value from the query result in text format and then measuring the
size of the buffer before resetting it. The result is used to get the
RU consumption with the tenant cost config's `PGWireEgressCost` method.

**sql: surface query request units consumed due to cpu usage**

This commit adds the ability for clients to estimate the number of RUs
consumed by a query due to CPU usage. This is accomplished by keeping a
moving average of the CPU usage for the entire tenant process, then using
that to obtain an estimate for what the CPU usage *would* be if the query
wasn't running. This is then compared against the actual measured CPU usage
during the query's execution to get the estimate. For local flows this is
done at the `connExecutor` level; for remote flows this is handled by the
last outbox on the node (which gathers and sends the flow's metadata).
The resulting RU estimate is added to the existing estimate from network
egress and displayed in the output of `EXPLAIN ANALYZE`.

**sql: surface query request units consumed by IO**

This commit adds tracking for request units consumed by IO operations
for all execution operators that perform KV operations. The corresponding
RU count is recorded in the span and later aggregated with the RU consumption
due to network egress and CPU usage. The resulting query RU consumption
estimate is visible in the output of `EXPLAIN ANALYZE`.

**multitenantccl: add sanity testing for ru estimation**

This commit adds a sanity test for the RU estimates produced by running
queries with `EXPLAIN ANALYZE` on a tenant. The test runs each test query
several times with `EXPLAIN ANALYZE`, then runs all test queries without
`EXPLAIN ANALYZE` and compares the resulting actual RU measurement to the
aggregated estimates. For now, this test is disabled during builds because
it is flaky in the presence of background activity. For this reason it
should only be used as a manual sanity test.

Informs #74441

Release note (sql change): Added an estimate for the number of request units
consumed by a query to the output of `EXPLAIN ANALYZE` for tenant sessions.

DrewKimball commented:
Alright, I made a few changes that should hopefully tie up the last few loose ends for this PR.
1. The network RU consumption estimation is no longer handled by the query's DistSQLReceiver; instead there is a pair of functions that hook into pgwire code that buffers the data, then measures its size and discards it.
2. The `planNodeToRowSource` processor now propagates RU consumption for IOs. This allows RUs to be measured for mutations, e.g. inserts and deletes.
3. The sanity test is now skipped to ensure it is only run manually. This is necessary because the test is very vulnerable to background activity, and will eventually flake if run as part of the test builds even if generous allowances are used.

I'm leaving two items for a future PR(s) for the sake of decreasing the burden of reviewing:
1. RU consumption estimation for external IOs / non-vectorized plans.
2. Adding a cluster setting to disable the RU estimation logic to give an escape hatch for the backport.

yuzefovich reviewed (approved):
I have some nits, but nothing blocking.

Curious though what prompted to make the change number 1?

> Reviewed 33 of 33 files at r24, 21 of 21 files at r25, 16 of 16 files at r26, 3 of 3 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-:-NH2kcLh-Nuv3LDZF-0I:bvqagb6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/plan_node_to_row_source.go` line 160 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2ixzhD1uAJLyZ2m_-:-NH2ixzhD1uAJLyZ2m_0:b-e1ewf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/978e6ef652db25e1aef2f4ac187606ceb96b249d/pkg/sql/plan_node_to_row_source.go#L160)):*
> ```Go
> 	ctx = p.StartInternal(ctx, nodeName(p.node))
> 	p.params.ctx = ctx
> 	if execstats.ShouldCollectStats(ctx, p.FlowCtx.CollectStats) {
> ```

nit: let's move this to occur earlier. All other processors do this in their constructor, and I think we could do it in `newPlanNodeToRowSource` too. If not, after `InitWithEvalCtx` in `Init` should work.
___
*[`pkg/sql/plan_node_to_row_source.go` line 258 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2jpza21w5o3xjJcL9:-NH2jpza21w5o3xjJcLA:b-p9yw1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/978e6ef652db25e1aef2f4ac187606ceb96b249d/pkg/sql/plan_node_to_row_source.go#L258)):*
> ```Go
> 	scanStats := execstats.GetScanStats(p.Ctx, p.ExecStatsTrace)
> 	if scanStats.ConsumedRU == 0 {
> 		return nil
> ```

nit: what would happen if we returned a non-nil object with ConsumedRU of 0? I'd guess it'd be ignored later down the line. This would allow us to not add a non-nil check in the columnarizer. Up to you.
___
*[`pkg/sql/pgwire/conn.go` line 1418 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2giHyCICVTOGkYixm:-NH2giHyCICVTOGkYixn:b4yhg3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd3d0c7ece449eba5777b2b6ad47288e882626be/pkg/sql/pgwire/conn.go#L1418)):*
> ```Go
> 
> 	var conv sessiondatapb.DataConversionConfig
> 	buf := newWriteBuffer(nil /* byteCount */)
> ```

nit: we could reuse this `writeBuffer` between different invocations to save on some of the allocations. I'm thinking about introducing a struct that would hold on the write buffer and would export two methods (i.e. implement an interface defined in `sql` package). We'd probably then need to inject a constructor method into `sql` package. However, the write buffer already has some scratch space that the allocations might not happen, and this might not be a big deal after all. Up to you.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 33 of 33 files at r24, 21 of 21 files at r25, 16 of 16 files at r26, 3 of 3 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
I found that a NPE was possible when the row implementation was used because the `DistSQLReceiver` isn't set up with the column types of the queried rows for `EXPLAIN ANALYZE`. I could have changed the function interface to also pass a types slice, but it seemed safer to explicitly set everything up for the purpose of buffering and counting bytes to ensure no more nasty surprises.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-:-NH6WZFb31mwhrK6PJYc:bsm6w4j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @yuzefovich)

___
*[`pkg/sql/plan_node_to_row_source.go` line 160 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2ixzhD1uAJLyZ2m_-:-NH6Y4jeBDw4SoJ3HPwy:bxvzplr) ([raw file](https://github.com/cockroachdb/cockroach/blob/978e6ef652db25e1aef2f4ac187606ceb96b249d/pkg/sql/plan_node_to_row_source.go#L160)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: let's move this to occur earlier. All other processors do this in their constructor, and I think we could do it in `newPlanNodeToRowSource` too. If not, after `InitWithEvalCtx` in `Init` should work.
</blockquote></details>

Moved it to `InitWithOutput`, since we need the `FlowCtx`. Done.
___
*[`pkg/sql/plan_node_to_row_source.go` line 258 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2jpza21w5o3xjJcL9:-NH6VUNM3rrHMQLTKd5G:b3nxbxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/978e6ef652db25e1aef2f4ac187606ceb96b249d/pkg/sql/plan_node_to_row_source.go#L258)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: what would happen if we returned a non-nil object with ConsumedRU of 0? I'd guess it'd be ignored later down the line. This would allow us to not add a non-nil check in the columnarizer. Up to you.
</blockquote></details>

Yeah that would be fine, but the comment for `ProcessorBaseNoHelper.ExecStatsForTrace` says it can return nil, so it seems like we should do this anyway. Is there something special about the columnarizer so that it doesn't have to worry about that?
___
*[`pkg/sql/pgwire/conn.go` line 1418 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2giHyCICVTOGkYixm:-NH6YZaC2xHcltKVLusm:b7gm1sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd3d0c7ece449eba5777b2b6ad47288e882626be/pkg/sql/pgwire/conn.go#L1418)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we could reuse this `writeBuffer` between different invocations to save on some of the allocations. I'm thinking about introducing a struct that would hold on the write buffer and would export two methods (i.e. implement an interface defined in `sql` package). We'd probably then need to inject a constructor method into `sql` package. However, the write buffer already has some scratch space that the allocations might not happen, and this might not be a big deal after all. Up to you.
</blockquote></details>

I like that idea, won't be significantly more complex and will save allocations. Done.
___
*[`pkg/sql/rowexec/inverted_joiner.go` line 782 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NFoyFYn8SqPvEniPjy0:-NH6WUHM7OM_HOqcfmLL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/696a117a3c9e76d5d6e09cda5f92e7518d776e9c/pkg/sql/rowexec/inverted_joiner.go#L782)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Note that we don't calculate RU for non-tenants today, so `ConsumedRU` should be zero in those cases.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 7 of 7 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89256)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/plan_node_to_row_source.go` line 258 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/89256#-NH2jpza21w5o3xjJcL9:-NH74w7fESFjhvsMsIMV:b-2yzgpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/978e6ef652db25e1aef2f4ac187606ceb96b249d/pkg/sql/plan_node_to_row_source.go#L258)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Yeah that would be fine, but the comment for `ProcessorBaseNoHelper.ExecStatsForTrace` says it can return nil, so it seems like we should do this anyway. Is there something special about the columnarizer so that it doesn't have to worry about that?
</blockquote></details>

Oh yeah, you're right, let's keep the nil check in. For some reason I assumed that the meaning of "Can return nil" was equivalent to "This field can be nil".


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

DrewKimball commented:
bors r+

DrewKimball commented:
Here's an example of what the output looks like:
```
root@localhost:26269/defaultdb> CREATE TABLE xy (x INT, y INT);

CREATE TABLE


Time: 133ms total (execution 132ms / network 1ms)

root@localhost:26269/defaultdb> EXPLAIN ANALYZE INSERT INTO xy (SELECT t, t%27 FROM generate_series(1, 10000) g(t));

                 info
--------------------------------------
  planning time: 119ms
  execution time: 256ms
  distribution: local
  vectorized: true
  maximum memory usage: 100 KiB
  network usage: 0 B (0 messages)
  estimated RUs consumed: 20,553

  • insert
  │ nodes: n1
  │ actual row count: 1
  │ into: xy(x, y, rowid)
  │ auto commit
  │
  └── • render
      │
      └── • project set
          │ nodes: n1
          │ actual row count: 10,000
          │ estimated row count: 10
          │
          └── • emptyrow
                nodes: n1
                actual row count: 1
(24 rows)


Time: 376ms total (execution 376ms / network 0ms)

root@localhost:26269/defaultdb> EXPLAIN ANALYZE SELECT * FROM xy;

                        info
-----------------------------------------------------
  planning time: 293µs
  execution time: 110ms
  distribution: full
  vectorized: true
  rows read from KV: 10,000 (412 KiB, 1 gRPC calls)
  cumulative time spent in KV: 107ms
  maximum memory usage: 460 KiB
  network usage: 0 B (0 messages)
  estimated RUs consumed: 187

  • scan
    nodes: n1
    actual row count: 10,000
    KV time: 107ms
    KV contention time: 0µs
    KV rows read: 10,000
    KV bytes read: 412 KiB
    KV gRPC calls: 1
    estimated max memory allocated: 450 KiB
    missing stats
    table: xy@xy_pkey
    spans: FULL SCAN
(22 rows)


Time: 111ms total (execution 111ms / network 1ms)

root@localhost:26269/defaultdb> EXPLAIN ANALYZE SELECT count(*) FROM xy;

                        info
-----------------------------------------------------
  planning time: 295µs
  execution time: 4ms
  distribution: full
  vectorized: true
  rows read from KV: 10,000 (412 KiB, 1 gRPC calls)
  cumulative time spent in KV: 4ms
  maximum memory usage: 440 KiB
  network usage: 0 B (0 messages)
  estimated RUs consumed: 7

  • group (scalar)
  │ nodes: n1
  │ actual row count: 1
  │
  └── • scan
        nodes: n1
        actual row count: 10,000
        KV time: 4ms
        KV contention time: 0µs
        KV rows read: 10,000
        KV bytes read: 412 KiB
        KV gRPC calls: 1
        estimated max memory allocated: 430 KiB
        missing stats
        table: xy@xy_pkey
        spans: FULL SCAN
(26 rows)


Time: 5ms total (execution 5ms / network 0ms)
```
